### PR TITLE
feat: validate reviewer inline anchors against diff ranges

### LINKED ISSUE #127: feat: validate reviewer inline anchors against diff ranges
## Summary
Build a diff anchor index for reviewer prompts and publishing so comments are easier to anchor inline and invalid anchors can be detected or downgraded safely.

## Scope
- Parse the unified PR diff into anchorable ranges by file, side, and line range.
- Include an `ANCHORABLE DIFF LOCATIONS` section in the reviewer prompt with concise path/range/excerpt context.
- Validate reviewer-provided inline `path`, `line`, `side`, `startLine`, and `startSide` against the anchor index.
- Preserve exact valid anchors.
- Safely downgrade invalid anchors to top-level comments with location context.
- For markdown/spec files, include nearby heading context when available.

## Acceptance criteria
- Unit tests cover single-file and multi-hunk diffs.
- Markdown heading context is captured for docs/spec files.
- Exact valid anchors are preserved.
- Out-of-range anchors are downgraded with clear fallback location text.
- Top-level comments without any location context are flagged for follow-up quality gating.

## Notes
This should come after the structured comment schema/rubric PR and can build on observe-mode quality gates.

### PR DESCRIPTION
**Pull request by mrcfps**
## Summary
- Add a diff anchor index that parses unified PR diffs into file/side line ranges with excerpt and Markdown/spec heading context.
- Include anchorable diff locations in reviewer prompts so agents can choose valid inline anchors.
- Normalize review anchors before publishing, preserving valid inline comments and downgrading invalid anchors into top-level feedback with fallback location context.

## Validation
- `go test ./...`
- `go vet ./...`
- `go build ./...`

Closes #127

<sub>Generated by looper 0.0.0-dev · runner=worker · agent=opencode</sub>

### CODE REVIEW

File: internal/reviewer/runner.go [hunk 1]
```diff
@@ -1685,6 +1686,7 @@ func buildReviewPrompt(repo string, prNumber int64, checkpoint reviewerCheckpoin
 		}
 		payload := parseJSONObject(optionalString(checkpoint.Snapshot.PayloadJSON))
 		if diff, ok := stringFromAny(payload["diff"]); ok && diff != "" {
+			parts = append(parts, diffanchor.Parse(diff).FormatPromptSection(80))
```
Comment (mrcfps) [inline internal/reviewer/runner.go:2198]: `buildReviewPrompt` limits `FormatPromptSection` to 80 ranges, but the generated section tells the reviewer to "Use only these path/side/line ranges". On any PR with more than 80 anchorable ranges, valid diff positions after the cutoff are presented as if they are unavailable, so a code-level finding there may be incorrectly downgraded into the top-level body instead of becoming a resolvable inline thread. Please either include all ranges, change the omitted-ranges wording to say the full diff remains authoritative for anchors, or avoid the hard "only these" contract when the list is truncated.

**Review by mrcfps — COMMENTED** [review]
The diff anchor index is a useful direction, but I found a few correctness gaps where the new safeguards can either mislead reviewers on larger diffs or fail to sanitize anchors before GitHub sees them.


<sub>Generated by looper 0.0.0-dev · runner=reviewer · agent=opencode</sub>

File: internal/reviewer/runner.go [hunk 2]
```diff
@@ -1711,7 +1713,7 @@ func buildReviewPrompt(repo string, prNumber int64, checkpoint reviewerCheckpoin
...
 		"Every review body you post must include exactly one stable idempotency marker with id, head, and outcome fields: `<!-- looper:review id=... head=... outcome=clean|actionable -->`.",
 		reviewDisclosureInstruction(disclosureCfg, agentRuntime, agentModel),
-		"If your inline review API call is rejected because of an invalid anchor, retry once with corrected inline anchors. If code-anchored actionable findings still cannot be submitted as resolvable inline review comments, exit non-zero instead of moving them into the review body or creating separate duplicate PR comments.",
+		"Before posting, validate every inline review comment's `path`, `line`, `side`, `start_line`, and `start_side` against ANCHORABLE DIFF LOCATIONS. Preserve exact anchors that fit those ranges. If an otherwise useful comment is outside those ranges, safely downgrade it to top-level review body feedback that starts with clear fallback location text instead of submitting an invalid inline anchor.",
```
Comment (mrcfps) [inline internal/reviewer/runner.go]: This new safety rule is only prompt text for the spawned reviewer, while the same prompt still requires the agent to publish directly with `gh`; looper does not parse or submit the review for that path. That means the new `normalizeReviewAnchors` enforcement in `Gateway.SubmitReview` is bypassed for the actual reviewer flow being instructed here, so invalid anchors can still be posted if the agent misses the prompt-only validation. Please route reviewer publishing through an enforced wrapper/Gateway path that validates the payload before calling GitHub, or otherwise make the PR explicit that this path remains best-effort and add tests around the real publishing mechanism.

File: internal/infra/github/review_anchor.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+			}
+		}
+		return body, comments, flags
+	}
+	kept := make([]ReviewComment, 0, len(comments))
+	downgraded := []string{}
+	for _, comment := range comments {
+		result := anchors.Validate(diffanchor.Anchor{Path: comment.Path, Line: comment.Line, Side: comment.Side, StartLine: comment.StartLine, StartSide: comment.StartSide})
+		if result.Valid {
+			kept = append(kept, normalizeReviewCommentAnchor(comment))
```
Comment (mrcfps) [inline internal/infra/github/review_anchor.go:30]: `diffanchor.Validate` trims `comment.Path` before checking the index, but `normalizeReviewCommentAnchor` only canonicalizes the side before returning the kept comment. A comment with a valid path plus accidental leading/trailing whitespace can therefore pass validation and then be sent to GitHub with the untrimmed path, causing the review submission to fail even though normalization reported it as safe. Please trim `comment.Path` in `normalizeReviewCommentAnchor` (and add a regression test with whitespace around the path) so the submitted anchor is the same canonical value that was validated.

File: internal/diffanchor/diffanchor.go [hunk 4]
```diff
@@ -0,0 +1,422 @@
...
+		}
+	}
+	for _, line := range strings.Split(diff, "\n") {
+		if strings.HasPrefix(line, "diff --git ") {
+			flush()
+			path = gitDiffPath(line)
+			rightHeading, leftHeading = "", ""
+			continue
+		}
+		if strings.HasPrefix(line, "+++ ") {
```
Comment (mrcfps) [inline internal/diffanchor/diffanchor.go]: This header check runs for every line, including lines inside a hunk. An added source line whose content starts with `++ ` is encoded in a unified diff as `+++ ...`, so this branch treats it as a file header, potentially changes `path`, skips the line counter increment, and corrupts all subsequent anchor ranges. The same issue exists for removed content starting with `-- ` and the `--- ` skip below. Track whether parsing is currently inside a hunk and only interpret `+++`/`---` as file headers before hunk content; add regression tests for added `++ ...` and removed `-- ...` lines.

File: internal/infra/github/review_anchor.go [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+func reviewQualityGateApplies(event string, body string) bool {
+	if strings.EqualFold(strings.TrimSpace(event), "REQUEST_CHANGES") {
+		return true
+	}
+	body = strings.ToLower(body)
+	if strings.Contains(body, "outcome=actionable") {
+		return true
+	}
+	if strings.Contains(body, "outcome=clean") || strings.EqualFold(strings.TrimSpace(event), "APPROVE") {
```
Comment (mrcfps) [inline internal/infra/github/review_anchor.go]: The quality gate can be disabled by any occurrence of the substring `outcome=clean`, or by any `APPROVE` event, without validating that there is exactly one well-formed Looper marker whose outcome matches the event. That means a malformed/actionable body can bypass the missing-location checks simply by mentioning `outcome=clean` in prose or by being submitted as an approval. Parse the actual idempotency marker and reject event/outcome mismatches instead of using substring checks.

File: internal/cliapp/review_submit.go [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+func (r *commandRuntime) reviewSubmit(cmd *cobra.Command, args []string) error {
+	repo, prNumber, err := parsePullRequestRef(args[0])
+	if err != nil {
+		return err
+	}
+	event := strings.ToUpper(strings.TrimSpace(getStringFlag(cmd, "event")))
+	if event == "" {
+		return fmt.Errorf("review submit requires --event COMMENT, APPROVE, or REQUEST_CHANGES")
+	}
+	if event != "COMMENT" && event != "APPROVE" && event != "REQUEST_CHANGES" {
```
Comment (mrcfps) [inline internal/cliapp/review_submit.go]: `review submit` accepts `REQUEST_CHANGES`, but the reviewer publish verification path only recognizes COMMENT, and APPROVE when auto-approve is enabled, for agent-native reviews. If an agent successfully posts a `REQUEST_CHANGES` review through this wrapper, `FindReviewMarker` will not consider it a valid matching review, so the loop can retry/rerun and post duplicate feedback for the same head. Either reject `REQUEST_CHANGES` here for reviewer submissions, or thread REQUEST_CHANGES through the allowed review events and side-effect handling.

File: internal/reviewer/runner.go [hunk 7]
```diff
@@ -1702,16 +1710,22 @@ func buildReviewPrompt(repo string, prNumber int64, checkpoint reviewerCheckpoin
 	if manual {
 		reviewRequestInstruction = "This is a manual reviewer run, so a current-user review request is not required before posting."
 	}
+	githubOperationContract := fmt.Sprintf("GitHub operation contract: submit exactly one PR review for this run through the trusted Looper CLI at `%s review submit %s#%d --event COMMENT|APPROVE --commit-id %s`, with the review JSON on stdin. The wrapper validates inline anchors against the live PR diff before it calls GitHub; do not use PATH-based `looper`, repository-local `go run ./cmd/looper`, `gh api repos/%s/pulls/%d/reviews`, or `gh pr review` directly for the review submission.", looperCLICommand, repo, prNumber, snapshotHeadSHA(checkpoint), repo, prNumber)
```
Comment (mrcfps) [inline internal/reviewer/runner.go]: This command template always advertises `--event COMMENT|APPROVE`, even when `allowApprove` is false and the prompt later says "Do not approve". The operation contract is the primary submission instruction, so agents can follow it and submit an approval that the loop then ignores as non-idempotent when approvals are disabled. Build the event portion from `allowApprove` (`COMMENT` only when false, `COMMENT|APPROVE` when true) so the trusted-wrapper instruction and marker verification rules stay aligned.

**Review by mrcfps — COMMENTED** [review]
I found a few correctness gaps in the new validation path that can still publish unsafe reviews or reject valid diff anchors.


<sub>Generated by looper 0.0.0-dev · runner=reviewer · agent=opencode</sub>

**Review by mrcfps — COMMENTED** [review]
I found a few correctness gaps in the new enforced review-submit path that can still publish reviews the runner cannot safely track, or can mark reviews clean while carrying actionable feedback.


<sub>Generated by looper 0.0.0-dev · runner=reviewer · agent=opencode</sub>

File: internal/cliapp/review_submit.go [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+	raw, err := io.ReadAll(cmd.InOrStdin())
+	if err != nil {
+		return fmt.Errorf("read review payload from stdin: %w", err)
+	}
+	var payload reviewSubmitPayload
+	if err := json.Unmarshal(raw, &payload); err != nil {
+		return fmt.Errorf("parse review payload JSON from stdin: %w", err)
+	}
+
+	loaded, err := r.loadConfig()
```
Comment (mrcfps) [inline internal/cliapp/review_submit.go:54]: The wrapper validates the requested event before loading config, but it never checks `defaults.allowAutoApprove` before accepting `--event APPROVE`. The reviewer prompt and the marker verification path both treat approvals as disabled when that setting is false, so an agent mistake can still publish an APPROVED review through this trusted wrapper and then leave the loop unable to recognize that approval as an allowed terminal marker. Please reject `APPROVE` here unless `loaded.Config.Defaults.AllowAutoApprove` is true, and add a regression test for the CLI path.

File: internal/cliapp/review_submit.go [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+		return fmt.Errorf("fetch PR diff for anchor validation: %w", err)
+	}
+	parsedAnchors := diffanchor.Parse(diff)
+	anchors = &parsedAnchors
+
+	comments := make([]githubinfra.ReviewComment, 0, len(payload.Comments))
+	for _, comment := range payload.Comments {
+		comments = append(comments, githubinfra.ReviewComment{Body: comment.Body, Path: comment.Path, Line: comment.Line, Side: comment.Side, StartLine: comment.StartLine, StartSide: comment.StartSide})
+	}
+	if err := gh.SubmitReview(cmd.Context(), githubinfra.SubmitReviewInput{Repo: repo, PRNumber: prNumber, Event: event, Body: payload.Body, CommitID: commitID, Comments: comments, Anchors: anchors, CWD: cwd}); err != nil {
```
Comment (mrcfps) [inline internal/cliapp/review_submit.go:95]: `review submit` validates the live head SHA, but it does not require the stdin body to contain exactly one well-formed `looper:review` marker whose `head=` matches `--commit-id`. A payload with no marker, multiple markers, or a stale marker can still be submitted if the top-level text has location context; the runner later searches for the exact id/head marker and will not treat that review as satisfying idempotency, which can produce duplicate reviews on retries. Please validate the body in the wrapper before calling `SubmitReview`: exactly one marker, matching head, and outcome/event compatibility.

File: internal/infra/github/gateway.go [hunk 10]
```diff
@@ -511,33 +525,44 @@ func (g *Gateway) GetPullRequestDiff(ctx context.Context, input GetPullRequestDi
...
 
 func (g *Gateway) SubmitReview(ctx context.Context, input SubmitReviewInput) error {
-	if len(input.Comments) > 0 {
+	var flags []reviewQualityFlag
+	input.Body, input.Comments, flags = normalizeReviewAnchors(input.Body, input.Comments, input.Anchors)
+	gateApplies, err := reviewQualityGateApplies(input.Event, input.Body)
+	if err != nil {
+		return err
+	}
+	if gateApplies && len(flags) > 0 {
```
Comment (mrcfps) [inline internal/infra/github/gateway.go:540]: A clean marker disables the quality gate even when `input.Comments` still contains review comments. With `outcome=clean`, `reviewQualityGateApplies` returns false, so this path skips any flags and serializes the comments anyway; downstream, the runner will treat the marker as a clean result and add the +1/spec-ready side effects despite the review containing actionable inline or downgraded feedback. Please make `outcome=clean` mutually exclusive with review comments/downgraded comments, or return the parsed outcome and reject clean reviews whenever comments were supplied.

File: internal/infra/github/review_anchor.go [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+}
+
+func reviewQualityGateApplies(event string, body string) (bool, error) {
+	event = strings.ToUpper(strings.TrimSpace(event))
+	markers := parseReviewIdempotencyMarkers(body)
+	markerComments := reviewMarkerRE.FindAllStringSubmatch(body, -1)
+	if len(markerComments) == 0 {
+		return true, nil
+	}
+	if len(markers) != 1 {
```
Comment (mrcfps) [inline internal/infra/github/review_anchor.go]: This check counts only well-formed parsed markers, not all `<!-- looper:review ... -->` marker comments. A body with one valid marker plus an extra malformed looper review marker still has `len(markers) == 1` and passes, even though the contract requires exactly one stable marker and the extra malformed marker can confuse idempotency parsing/debugging later. Please require both `len(markerComments) == 1` and `len(markers) == 1` before accepting the body.

File: internal/infra/github/review_anchor.go [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+		if marker.Outcome != "actionable" {
+			return true, fmt.Errorf("review marker outcome=%s does not match REQUEST_CHANGES event", marker.Outcome)
+		}
+	}
+	return marker.Outcome != "clean", nil
+}
+
+func normalizeReviewCommentAnchor(comment ReviewComment) ReviewComment {
+	comment.Side = normalizeReviewCommentSide(comment.Side)
+	if comment.StartLine <= 0 {
```
Comment (mrcfps) [inline internal/infra/github/review_anchor.go]: Single-line comments with `start_line == line` are considered valid by the diff-anchor validator and then forwarded with `start_line`/`start_side` still set. The prompt tells agents to omit those fields for single-line comments, and GitHub treats `start_line` as the beginning of a multi-line range; forwarding the redundant range can turn an otherwise valid single-line anchor into an API rejection after validation succeeded. Please normalize `StartLine == Line` by clearing `StartLine` and `StartSide`, and add a regression test for that payload shape.
